slick/src/main/scala/slick/jdbc/DriverDataSource.scala

```diff
@@ -46,42 +46,45 @@ class DriverDataSource(
   // State that gets initialized by `init`
   @volatile private[this] var registered: Boolean = false
   @volatile private[this] var initialized = false
-  private[this] var driver: Driver = _
-  private[this] var connectionProps: Properties = _
+  @volatile private[this] var driver: Driver = _
+  @volatile private[this] var connectionProps: Properties = _
```

Comment:
That would have been my intuitive understanding before working on all the multi-threaded stuff for Slick 3.0, too, but it's not how the Java Memory Model works.

In a nutshell (if I still remember everything correctly): The basic principle is the happens-before relationship. If A happens-before B, any value written by A will be seen by B. It does not matter at all for this purpose whether an individual variable is volatile or not. A happens-before relationship is established by a) code running sequentially in the same thread, b) writing and subsequently reading the same volatile variable, c) leaving and subsequently entering a synchronized block on the same monitor.

The interesting part for lock-free transfer of mutable state between threads is b). *All* state changes performed by thread A can be made visible to thread B by writing to a single volatile (or atomic) var at the end in A and reading the same var first in B. Slick relies on this for the implementation of the asynchronous execution engine.

I hope this explains my reasoning in the previous comment that reordering the writes would make it safe without the extra volatile vars. More volatile writes may be slower but in this context the difference should be negligible.

I think we can safely merge this if you retarget it to 3.1. If you'd like to fix `getParentLogger`, too, that would be even nicer, of course :-)